### PR TITLE
Add linktolatest option for RollingFileWriter

### PR DESCRIPTION
**Pull request by pabl0rg**
see https://github.com/pmwmedia/tinylog/issues/149

I'm not sure how to implement the assertions in the test.  The build is not working on my machine (javac 11.0.7)

```
tinylog/tinylog-api/src/main/java/org/tinylog/runtime/LegacyJavaRuntime.java:[58,43] cannot find symbol
[ERROR]   symbol:   class Reflection
[ERROR]   location: package sun.reflect
```

If you find the idea acceptable, could you please help me with this part?  Thank you!

### CODE REVIEW

**Comment by pmwmedia** [comment]
Thank you very much for your pull request :) You need Java 9 to compile all tinylog artifacts. Java 11 and later don't include `sun.reflect.Reflection` anymore in favor to the new stack walker API. Java 9 and 10 are the only Java versions that can compile the legacy reflection API as well as the new stack walker API. tinylog supports both and decides on runtime which to use.

Is there any way to implement it without using `Files.createSymbolicLink()`? The problem is that Android added this class with API level 26, but some developers still have to support older Android versions.

**Comment by pabl0rg** [comment]
It wasn't as easy as i expected to find a jdk 9 installer and to remember how to do things in java 1.6 , but it finally builds.  Please let me know what you think.

**Comment by pmwmedia** [comment]
Today, I read about creating symbolic links in Java without using the NIO API. Unfortunately, there is really no native Java API for it prior Java 7. Creating a process for executing OS commands as you did is the only way. However, the disadvantage is that these commands are OS depending. For example, it won't work on Windows.

However, I think your feature is very useful for computer applications, but I don't see any use case for Android apps. Therefore, I prefer to use your previous approach with the `Files` class from the NIO API. You could check if tinylog is running on Android and log an error message. Otherwise, you can execute your code. I have just added a method to detect whether running on Android: `RuntimeProvider.isAndroid()`

**Comment by pmwmedia** [comment]
By the way, I have a question about `removeLinkToLatest()`: Why do you delete the symbolic link?

**Comment by pabl0rg** [comment]
I renamed `removeLinkToLatest` to `filterOutSymlinks`.  It exists in case the user puts the symlink in the same folder as the actual log files.  The logic seems much easier to read now with the java 7 sdk.  

However, I am having trouble getting the build to succeed (it seems to exclude java 7 sdk):
```
[ERROR] /Users/juanliska/Documents/startrack/code/tinylog/tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java:129: Undefined reference: java.nio.file.Path
[ERROR] /Users/juanliska/Documents/startrack/code/tinylog/tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java:129: Undefined reference: java.nio.file.Path java.nio.file.Files.createSymbolicLink(java.nio.file.Path, java.nio.file.Path, java.nio.file.attribute.FileAttribute[])
```

File: tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java [hunk 1]
```diff
@@ -96,7 +100,39 @@ public RollingFileWriter(final Map<String, String> properties) throws FileNotFou
 		charset = getCharset(properties);
 		buffered = Boolean.parseBoolean(properties.get("buffered"));
 		writingThread = Boolean.parseBoolean(properties.get("writingthread"));
-		writer = createByteArrayWriter(fileName, append, buffered, false, false);
+		writer = createByteArrayWriterAndLinkLatest(fileName, append, buffered, false, false);
+	}
+
+	private static List<File> filterOutSymlinks(final List<File> files) {
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java:109]: You can add `@IgnoreJRERequirement` to allow JRE 7 API for this method

File: tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java [hunk 2]
```diff
@@ -96,7 +100,39 @@ public RollingFileWriter(final Map<String, String> properties) throws FileNotFou
...
+				if (java.nio.file.Files.isSymbolicLink(file.toPath())) {
+					symlinks.add(file);
+				}
+			}
+			files.removeAll(symlinks);
+		}
+		return files;
+	}
+
+	private ByteArrayWriter createByteArrayWriterAndLinkLatest(final String fileName, final boolean append, final boolean buffered,
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java:123]: You can add `@IgnoreJRERequirement` to allow JRE 7 API also for this method

File: tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java [hunk 3]
```diff
@@ -96,7 +100,39 @@ public RollingFileWriter(final Map<String, String> properties) throws FileNotFou
...
+		final boolean threadSafe, final boolean shared) throws FileNotFoundException {
+		ByteArrayWriter writer = AbstractFormatPatternWriter.createByteArrayWriter(fileName, append, buffered, threadSafe, shared);
+		if (linkToLatest != null) {
+			File logFile = new File(fileName);
+			File linkFile = new File(linkToLatest.resolve());
+			if (!RuntimeProvider.isAndroid()) {
+				try {
+					Path linkPath = linkFile.toPath();
+					java.nio.file.Files.delete(linkPath);
+					java.nio.file.Files.createSymbolicLink(linkPath, logFile.toPath());
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java]: Please add an import statement for `java.nio.file.Files` to make it more readable

File: tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java [hunk 4]
```diff
@@ -96,7 +100,39 @@ public RollingFileWriter(final Map<String, String> properties) throws FileNotFou
...
+		return files;
+	}
+
+	private ByteArrayWriter createByteArrayWriterAndLinkLatest(final String fileName, final boolean append, final boolean buffered,
+		final boolean threadSafe, final boolean shared) throws FileNotFoundException {
+		ByteArrayWriter writer = AbstractFormatPatternWriter.createByteArrayWriter(fileName, append, buffered, threadSafe, shared);
+		if (linkToLatest != null) {
+			File logFile = new File(fileName);
+			File linkFile = new File(linkToLatest.resolve());
+			if (!RuntimeProvider.isAndroid()) {
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java:129]: Could you issue an internal logger warning for Android?

File: tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java [hunk 5]
```diff
@@ -96,7 +100,39 @@ public RollingFileWriter(final Map<String, String> properties) throws FileNotFou
...
+				}
+			}
+			files.removeAll(symlinks);
+		}
+		return files;
+	}
+
+	private ByteArrayWriter createByteArrayWriterAndLinkLatest(final String fileName, final boolean append, final boolean buffered,
+		final boolean threadSafe, final boolean shared) throws FileNotFoundException {
+		ByteArrayWriter writer = AbstractFormatPatternWriter.createByteArrayWriter(fileName, append, buffered, threadSafe, shared);
```
Comment (pmwmedia) [inline tinylog-impl/src/main/java/org/tinylog/writers/RollingFileWriter.java]: The class name `AbstractFormatPatternWriter` is redundant here, you can remove it to make the line a bit shorter

**Review by pmwmedia — CHANGES_REQUESTED** [review]
Thank you very much for your awesome pull request! I have just reviewed it. I have only a few minor change requests. Could you apply them? Afterwards I'm going to merge your pull request.

**Comment by pabl0rg** [comment]
Thanks for your useful comments!  I made the suggested changes.

**Comment by pmwmedia** [comment]
@pabl0rg Can you reproduce the failed JUnit test on your local machine?

**Comment by pabl0rg** [comment]
Actually it seems that test case was not actually needed/wanted. (symlinks should only updated on rollover)

**Comment by pmwmedia** [comment]
@pabl0rg I have merged your pull request! Thank you very much for your contribution :)

I just noticed that you used an email address (jp@s***.info) for your commits that isn't linked to your GitHub profile. Could you add it (https://github.com/settings/emails)?

**Comment by pabl0rg** [comment]
Done!  Thank you and sorry I didn't fix the tests.  I came down with covid and my thinking has been foggy.

**Comment by pmwmedia** [comment]
@pabl0rg Thank you very much! Health has first priority of course. I hope you feel already better. Get well soon!

**Pull request merged**